### PR TITLE
All the changes I promised - and more

### PR DESCRIPTION
**Pull request by dansarie**
The most important changes are:

Fix indentation and trailing whitespace in XML source.

Changes to structure and language to make the draft easier
to follow for implementers.

The most significant 3 bytes of the timestamp contain the MJD. This
leaves 5 bytes for the number of microseconds since midnight, which is
necessary since 4 aren't enough to represent all 86,400,000,000
microseconds in a day.

Added signature contexts, which are used by all servers but were never
documented.

Relaxed the requirement that requests MUST be at least 1024 bytes to
SHOULD, in order to enable clients on network with low MTUs to use the
protocol. Servers still MUST NOT sent responses longer than the requests.

Allow ASCII tags to be shorter than four characters and auto-pad to four
bytes in the tag representation. Renamed PAD\xff to PAD, which will be
backward compatible since servers MUST ignore unknown tags according to
the earlier draft.

### CODE REVIEW

**Review by aanchal4 — APPROVED** [review]

**Review by aanchal4 — APPROVED** [review]

**Review by aanchal4 — APPROVED** [review]

**Review by aanchal4 — CHANGES_REQUESTED** [review]
"The server incorporates the nonce into its signed response so that the client can verify that the response was created after the nonce. Provided that the nonce has sufficient entropy, this proves that the signed response could only have been generated after the nonce."
SHOULD BE
"The server incorporates the nonce into its signed response so that the client can verify server's signatures with the nonce issued by the client. Provided that the nonce has sufficient entropy, this proves that the signed response could only have been generated after the nonce."

**Review by aanchal4 — APPROVED** [review]

**Review by aanchal4 — CHANGES_REQUESTED** [review]
These may be represented as an array of objects in JSON <xref target="RFC8259"/> where each object may have keys "blind" and "packet". Packet has the Base64 <xref target="RFC4648"/> encoded bytes of the packet and blind is the blind used for the next nonce.
I think it should be the following:
These may be represented as an array of objects in JSON <xref target="RFC8259"/> where each object may have keys "blind" and "response packet". Packet has the Base64 <xref target="RFC4648"/> encoded bytes of the last response packet and blind is the blind used for the next nonce.

**Review by aanchal4 — CHANGES_REQUESTED** [review]
"Tags MUST be listed in the same order as their corresponding offsets."
I think the following statement is more clear:
"Tags MUST be listed in the same order as the offsets of their values."

**Review by aanchal4 — CHANGES_REQUESTED** [review]
1) "..... servers which sent signed time responses in return."
should be
"..... servers who send signed time responses in return."

2) "...and servers MUST NOT send responses larger than the requests it is replying to."
should be 
""...and servers MUST NOT send responses larger than the requests they are replying to.""

3) "It SHOULD be generated by the client either at random or by hashing a previous Roughtime response message together with a blind as described in <xref target="cheater-detection"/>."

If the value for NONC is generated at random, the cheater detection will not work. right?
Instead we must say that blind should be chosen randomly.

4) "The nonce MUST NOT be predictable by an outside observer"
"MUST NOT" in this sentence is a bit odd. This is not an enforceable statement.

5) "The SIG tag value is a 64 byte Ed25519 signature <xref target="RFC8032"/> over the entire value of another tag."
should be:
"The SIG tag value is a 64 byte Ed25519 signature <xref target="RFC8032"/> over the entire value of SREP tag."

6) "The CERT tag contains a signing key certificate signed with a long term key." 
The term "signing key certificate" is non-standard. I think it should be: 
"The CERT tag contains a public-key certificate signed with the server's long term key." 

7) "The DELE tag contains a delegated signing certificate used by the server to sign the SREP tag."
Should be
"The DELE tag contains a delegated public-key certificate used by the server to sign the SREP tag."

**Comment by dansarie** [comment]
> "The server incorporates the nonce into its signed response so that the client can verify that the response was created after the nonce. Provided that the nonce has sufficient entropy, this proves that the signed response could only have been generated after the nonce."
> SHOULD BE
> "The server incorporates the nonce into its signed response so that the client can verify server's signatures with the nonce issued by the client. Provided that the nonce has sufficient entropy, this proves that the signed response could only have been generated after the nonce."

I don't understand this. The client doesn't use the nonce to verify the server's signatures, but that the signed response was generated after the nonce.

> These may be represented as an array of objects in JSON where each object may have keys "blind" and "packet". Packet has the Base64 encoded bytes of the packet and blind is the blind used for the next nonce.
> I think it should be the following:
> These may be represented as an array of objects in JSON where each object may have keys "blind" and "response packet". Packet has the Base64 encoded bytes of the last response packet and blind is the blind used for the next nonce.

Fixed in f1f48b3, but I replaced "response packet" with "response_packet" to comply with JSON key naming rules.

> "Tags MUST be listed in the same order as their corresponding offsets."
> I think the following statement is more clear:
> "Tags MUST be listed in the same order as the offsets of their values."

Agree, fixed in f1f48b3.

> 1. "..... servers which sent signed time responses in return."
>    should be
>    "..... servers who send signed time responses in return."

Fixed in f1f48b3.

> 2. "...and servers MUST NOT send responses larger than the requests it is replying to."
>    should be
>    ""...and servers MUST NOT send responses larger than the requests they are replying to.""

Fixed in f1f48b3.

> 3. "It SHOULD be generated by the client either at random or by hashing a previous Roughtime response message together with a blind as described in ."
> If the value for NONC is generated at random, the cheater detection will not work. right?
> Instead we must say that blind should be chosen randomly.

Right, that's why the sentence says "*either* at random *or* by hashing a previous Roughtime response".

> 4. "The nonce MUST NOT be predictable by an outside observer"
>    "MUST NOT" in this sentence is a bit odd. This is not an enforceable statement.

Agree, removed in f1f48b3.

> 5. "The SIG tag value is a 64 byte Ed25519 signature  over the entire value of another tag."
>    should be:
>    "The SIG tag value is a 64 byte Ed25519 signature  over the entire value of SREP tag."

This has already been fixed in 2bf445b2.

> 6. "The CERT tag contains a signing key certificate signed with a long term key."
>    The term "signing key certificate" is non-standard. I think it should be:
>    "The CERT tag contains a public-key certificate signed with the server's long term key."

Fixed in f1f48b3.

> 7. "The DELE tag contains a delegated signing certificate used by the server to sign the SREP tag."
>    Should be
>    "The DELE tag contains a delegated public-key certificate used by the server to sign the SREP tag."

Fixed in f1f48b3.

I've also fixed a number of other things I found after going over the draft again:
* d3c35c5f76e8166edf03a7174573c01a4df12561 changes the title of Section 1 to comply with RFC 7322.
* 7b7e6550b75d9f755edef1c3e994a7fb5feb2a66 makes SHOULD lowercase in the introduction.
* 17274d4d270b966628ad072eae1f5c4ded83798f contains a number of language fixes and clarifications.
* 39e9169014915f4ebe2ae50953606b23a7dd8866 changes "long term" to "long-term" when it is used as an adjective.
* 715bb31c43ae18e938fbaa69a8aadce7dc74d250 specifies that the UTC timescale should be used and includes a reference to its definition in ITU-R TF.460-6
* e14967d4a6a2e9fa037627139bf19d5cafa9eb02 moved three references to the normative section (see https://www.ietf.org/blog/iesg-statement-normative-and-informative-references/).
* 6911941625ac44600c9332fd2105437a8c2ce74f removes a superfluous "\x00".
* 991d7da7d3453635c08b83f93a8c7ea219476226 writes out the JSON abbreviation and adds it to the list of abbreviations.
* bf4f7c5976fa27589d5f0c5d8cbc4cbda8a1f1e6 adds a DOI to [DelayAttacks].
* 04be9a39c8b677797be43787bb2b5193d41df2d8 removes "value" from the definition of the tag registry to avoid confusion.
* 85ab6c8e25a0722a41ca52d446dcb2d0fdd7c8f1 Fixes a typo in a reference.

**Comment by aanchal4** [comment]
> I don't understand this. The client doesn't use the nonce to verify the server's signatures, but that the signed response was generated after the nonce.

Let me rephrase "The server incorporates the nonce into its signed response so that the client can verify server's signatures covering the nonce issued by the client. Provided that the nonce has sufficient entropy, this proves that the signed response could only have been generated after the nonce"

**Comment by aanchal4** [comment]
> Right, that's why the sentence says "_either_ at random _or_ by hashing a previous Roughtime response".

Why do we want to give the choice that defeats the main goal of the protocol? And if we do we should clearly specify the caveat.

**Comment by dansarie** [comment]
> Let me rephrase "The server incorporates the nonce into its signed response so that the client can verify server's signatures covering the nonce issued by the client. Provided that the nonce has sufficient entropy, this proves that the signed response could only have been generated after the nonce"

Fixed in d696dc225d12f050f7f216ea8dbc48b84830d41f.

> Why do we want to give the choice that defeats the main goal of the protocol? And if we do we should clearly specify the caveat.

I see your point now. My idea was that the client would generate the nonce at random only if it had no previous responses available. d696dc225d12f050f7f216ea8dbc48b84830d41f makes changes to emphasize this.

Another thing I see now is that the first and third paragraph of "Protocol Overview" say the same things to some extent - mostly due to text added by me. I propose we remove everything after "received the client's request." in the first paragraph and change the third paragraph to:
```
The server proves freshness of its response as follows: The client's request
contains a nonce. The server incorporates a value derived from the nonce into
its signed response so that the client can verify the server's signatures
covering the nonce issued by the client. Provided that the nonce has sufficient
entropy, this proves that the signed response could only have been generated
after the nonce.
```
We then add another paragraph with:
```
Amortization of signing operations: The derived value included in the server's
response is the root of a Merkle tree which includes the hash of the client's
nonce as the value of one of its leaf nodes. This allows the server to perform a
single signing operation to create verifiable responses to a number of client
requests.
```
